exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,624 @@
+
+    fn add_started_service(&mut self, instance: Instance) {
+        self.started_services_by_name
+            .insert(instance.name.clone(), instance.id);
+        self.started_services.insert(instance.id, instance);
+    }
+
+    fn remove_started_service(&mut self, id: &InstanceId) {
+        let service = self.started_services.remove(id);
+        if let Some(instance) = service {
+            self.started_services_by_name.remove(&instance.name);
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| match err.0 {
+            JniErrorKind::JavaException => Error::JavaException.into(),
+            _ => Error::UnspecifiedError.into(),
+        })
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(
+            ErrorKind::runtime(value as u8),
+            format!("{:?}", value.clone()),
+        )
+    }
+}
+
+impl Runtime for JavaRuntimeProxy {
+    fn deploy_artifact(
+        &mut self,
+        artifact: ArtifactId,
+        deploy_spec: Any,
+    ) -> Box<dyn Future<Item = (), Error = ExecutionError>> {
+
+        let id = match self.parse_artifact(&artifact) {
+            Ok(id) => id,
+            Err(err) => return Box::new(Err(err).into_future()),
+        };
+
+        if let Err(err) = self.can_deploy_artifact(&id) {
+            return Box::new(Err(err).into_future());
+        }
+
+        Box::new(Self::parse_jni(self.exec.with_attached(|env| {
+            let artifact_id = JObject::from(env.new_string(id.to_string())?);
+            let spec = JObject::from(env.byte_array_from_slice(&deploy_spec.into_bytes())?);
+
+            env.call_method_unchecked(
+                self.runtime_adapter.as_obj(),
+                runtime_adapter::deploy_artifact_id(),
+                JavaType::Primitive(Primitive::Void),
+                &[
+                    JValue::from(artifact_id),
+                    JValue::from(spec),
+                ],
+            )?;
+            Ok(())
+        })).map(|result| {
+            self.add_deployed_artifact(id);
+            result
+        })
+        .into_future())
+    }
+
+    fn artifact_protobuf_spec(&self, id: &ArtifactId) -> Option<ArtifactProtobufSpec> {
+        let id = self.parse_artifact(id).ok()?;
+
+        if self.deployed_artifacts.contains(&id) {
+            // TODO: call `ServiceRuntimeAdapter`
+            Some(ArtifactProtobufSpec::default())
+        } else {
+            None
+        }
+    }
+
+    fn start_service(&mut self, spec: &InstanceSpec) -> Result<(), ExecutionError> {
+        let adapter = self.runtime_adapter.as_obj().clone();
+        let service_name = spec.name.clone();
+        let id = spec.id;
+        let artifact = self.parse_artifact(&spec.artifact)?;
```

Comment:
Here we parse the artifactId, only to convert it to string later. It seems unnecessary, as the correctness of such ids is checked before deploy in Java (and also each time it is instantiated in Java).